src/org/sosy_lab/java_smt/basicimpl/IndependentInterpolatingProverEnvironment.java

```diff
@@ -0,0 +1,548 @@
+  }
+
+  /**
+   * Identifies the shared variables between two sets of {@link BooleanFormula}s A and B.
+   *
+   * @param varsOfA A list of all variables in formulas of A.
+   * @param varsOfB A list of all variables in formulas of B.
+   * @return An {@link ImmutableList} of all variables found in both sets of formulas A and B.
+   */
+  private ImmutableList<Formula> getSharedVars(List<Formula> varsOfA, List<Formula> varsOfB) {
+    return varsOfA.stream().filter(varsOfB::contains).collect(ImmutableList.toImmutableList());
+  }
+
+  /**
+   * Determines the variables bound to a quantifier in a quantified set of formulas used for the
+   * quantifier-based interpolation strategy. These variables appear only in one set of formulas
+   * (either A or B) and can be safely removed by quantifier elimination.
+   *
+   * @param vars A list of variables in the set of formulas to determine the bound variables.
+   * @param sharedVars A list of variables found in both sets of formulas A and B.
+   * @return An {@link ImmutableList} of bound variables, returns an empty list if these variables
+   *     appear in both sets of formulas A and B.
+   */
+  private ImmutableList<Formula> getBoundVars(List<Formula> vars, List<Formula> sharedVars) {
+    return vars.stream()
+        .filter(var -> !sharedVars.contains(var))
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  /**
+   * Declares and calls an uninterpreted function with a unique identifier to generate interpolants.
+   *
+   * @param sharedVars A list of variables found in both sets of formulas A and B.
+   * @return An uninterpreted function with the {@code sharedVars} as symbols
+   */
+  private BooleanFormula getUniqueInterpolant(ImmutableList<Formula> sharedVars) {
+    return ufmgr.declareAndCallUF(
+        PREFIX + UNIQUE_ID_GENERATOR.getFreshId(), FormulaType.BooleanType, sharedVars);
+  }
+
+  /**
+   * Create a new, distinct prover to interpolate on. The set of formulas for interpolation will
+   * need to be translated to and from this new {@link ProverEnvironment} instance.
+   *
+   * @return A new {@link ProverEnvironment} configured to generate models.
+   */
+  private ProverEnvironment getDistinctProver() {
+    return solverContext.newProverEnvironment(ProverOptions.GENERATE_MODELS);
+  }
+
+  /**
+   * Checks, whether the returned interpolant indeed satisfies Craig interpolation.
+   *
+   * @param itp The given interpolant to check if it satisfies the definition of Craig interpolants.
+   * @param formulasOfA The set of formulas A, combined into one {@link BooleanFormula}.
+   * @param formulasOfB The set of formulas B, combined into one {@link BooleanFormula}.
+   * @param varsOfA A list of all variables in formulas of A.
+   * @param varsOfB A list of all variables in formulas of B.
+   * @return {@code true}, if the given interpolant is a valid Craig interpolant, returns {@code
+   *     false} otherwise.
+   */
+  private boolean satisfiesInterpolationCriteria(
```

Comment:
This method does not check the criteria of an Craig-Interpolant as stated.

The criteria are:

-  the implication φ− ⇒ ψ holds,
- the conjunction ψ ∧ φ+ is unsatisfiable, and
- ψ only contains symbols that occur in both φ− and φ+.